### PR TITLE
OHRM5X-1010:  TIME - Reports - Employee Reports - Get All Operation

### PR DESCRIPTION
**Pull request by SandeepaOHRM**
This PR contains:

- API to get Employee Report Table headers
- API to get Employee Report Data with filters
- Unit testings for DAO Layer
- Integration testings for API Layer

### CODE REVIEW

File: symfony/plugins/orangehrmTimePlugin/Dao/TimesheetDao.php [hunk 1]
```diff
@@ -406,4 +408,95 @@ public function getDefaultTimesheet(
...
+            $q->setParameter('state', EmployeeReportsSearchFilterParams::TIMESHEET_STATE[1]);
+        }
+
+        return $this->getQueryBuilderWrapper($q);
+    }
+
+    /**
+     * @param  EmployeeReportsSearchFilterParams  $filterParams
+     * @return string
+     */
```
Comment (VageeshOrangeHRM) [inline symfony/plugins/orangehrmTimePlugin/Dao/TimesheetDao.php:494]: we can change the return type as int, then we no need to type caste inside the getMeta() method

File: symfony/plugins/orangehrmTimePlugin/Dto/EmployeeReportsSearchFilterParams.php [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+    public const INCLUDE_TIMESHEETS = [
+        self::INCLUDE_TIMESHEETS_APPROVED_ONLY,
+        self::INCLUDE_TIMESHEETS_ALL,
+    ];
+
+    public const TIMESHEET_STATE = [
+        'SUBMITTED',
+        'APPROVED',
+        'REJECTED'
+    ];
```
Comment (VageeshOrangeHRM) [inline symfony/plugins/orangehrmTimePlugin/Dto/EmployeeReportsSearchFilterParams.php:41]: As per @RajithaKumaras KT we can optimize this https://github.com/orangehrm/orangehrm/pull/1086/files#diff-bb31e6dc1dabbc89f27534ba496bfafc4bb4d36e76597add44ce25ddc6d43e0eR41

File: symfony/plugins/orangehrmTimePlugin/Dao/TimesheetDao.php [hunk 3]
```diff
@@ -406,4 +408,95 @@ public function getDefaultTimesheet(
...
+            $q->andWhere('timesheetItem.projectActivity = :activityId');
+            $q->setParameter('activityId', $filterParams->getActivityId());
+        }
+
+        //Timesheet items after fromDate (including fromDate)
+        if (!is_null($filterParams->getFromDate())) {
+            $q->andWhere($q->expr()->gte('timesheetItem.date', ':fromDate'));
+            $q->setParameter('fromDate', $filterParams->getFromDate());
+        }
+
```
Comment (VageeshOrangeHRM) [inline symfony/plugins/orangehrmTimePlugin/Dao/TimesheetDao.php:475]: we can refactor the code snippets related to the date range filter as if elseif statement
https://github.com/orangehrm/orangehrm/pull/1086/files#diff-eba82252347cc542b14f7b1251656b28681e538802c1f2cbf7d0e1ca70201879R286

File: symfony/plugins/orangehrmTimePlugin/Dao/TimesheetDao.php [hunk 4]
```diff
@@ -406,4 +408,95 @@ public function getDefaultTimesheet(
...
+        if (!is_null($filterParams->getActivityId())) {
+            $q->andWhere('timesheetItem.projectActivity = :activityId');
+            $q->setParameter('activityId', $filterParams->getActivityId());
+        }
+
+        //Timesheet items after fromDate (including fromDate)
+        if (!is_null($filterParams->getFromDate())) {
+            $q->andWhere($q->expr()->gte('timesheetItem.date', ':fromDate'));
```
Comment (VageeshOrangeHRM) [inline symfony/plugins/orangehrmTimePlugin/Dao/TimesheetDao.php:472]: we can refactor the code snippets related to the date range filter as if elseif statement
https://github.com/orangehrm/orangehrm/pull/1086/files#diff-eba82252347cc542b14f7b1251656b28681e538802c1f2cbf7d0e1ca70201879R286

File: symfony/plugins/orangehrmTimePlugin/Dao/TimesheetDao.php [hunk 5]
```diff
@@ -406,4 +408,95 @@ public function getDefaultTimesheet(
...
+            $q->setParameter('projectId', $filterParams->getProjectId());
+        }
+
+        if (!is_null($filterParams->getActivityId())) {
+            $q->andWhere('timesheetItem.projectActivity = :activityId');
+            $q->setParameter('activityId', $filterParams->getActivityId());
+        }
+
+        //Timesheet items after fromDate (including fromDate)
+        if (!is_null($filterParams->getFromDate())) {
```
Comment (VageeshOrangeHRM) [inline symfony/plugins/orangehrmTimePlugin/Dao/TimesheetDao.php:471]: we can refactor the code snippets related to the date range filter as if elseif statement
https://github.com/orangehrm/orangehrm/pull/1086/files#diff-eba82252347cc542b14f7b1251656b28681e538802c1f2cbf7d0e1ca70201879R286

File: symfony/plugins/orangehrmTimePlugin/Dto/EmployeeReportsSearchFilterParams.php (hunk 2 above)
Comment (SandeepaOHRM) [reply symfony/plugins/orangehrmTimePlugin/Dto/EmployeeReportsSearchFilterParams.php]: done

File: symfony/plugins/orangehrmTimePlugin/Dao/TimesheetDao.php (hunk 5 above)
Comment (SandeepaOHRM) [reply symfony/plugins/orangehrmTimePlugin/Dao/TimesheetDao.php]: done

File: symfony/plugins/orangehrmTimePlugin/Dao/TimesheetDao.php (hunk 1 above)
Comment (SandeepaOHRM) [reply symfony/plugins/orangehrmTimePlugin/Dao/TimesheetDao.php:501]: done

File: symfony/plugins/orangehrmTimePlugin/Api/TimeReportAPI.php [hunk 6]
```diff
@@ -29,7 +29,7 @@ class TimeReportAPI extends ReportAPI
 {
     public const TIME_REPORT_MAP = [
         'project_report' => ProjectReport::class,
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmTimePlugin/Api/TimeReportAPI.php:31]:
```suggestion
        'project' => ProjectReport::class,
```

**Review by RajithaKumara — CHANGES_REQUESTED** [review]

File: symfony/plugins/orangehrmTimePlugin/Report/EmployeeReport.php [hunk 7]
```diff
@@ -19,61 +19,208 @@
...
     /**
      * @inheritDoc
      */
     public function getValidationRule(EndpointProxy $endpoint): ParamRuleCollection
     {
-        // TODO: Implement getValidationRule() method.
+        return new ParamRuleCollection(
+            new ParamRule(
+                CommonParams::PARAMETER_EMP_NUMBER,
+                new Rule(Rules::ENTITY_ID_EXISTS, [Employee::class]),
```
[the remark this answers is not in the document]
Comment (SandeepaOHRM) [reply symfony/plugins/orangehrmTimePlugin/Report/EmployeeReport.php:118]: done

File: symfony/plugins/orangehrmTimePlugin/Report/EmployeeReport.php [hunk 8]
```diff
@@ -19,61 +19,208 @@
...
+                new Rule(Rules::IN_ACCESSIBLE_EMP_NUMBERS)
+            ),
+            $endpoint->getValidationDecorator()->notRequiredParamRule(
+                new ParamRule(
+                    self::FILTER_PARAMETER_PROJECT_ID,
+                    new Rule(Rules::POSITIVE)
```
[the remark this answers is not in the document]
Comment (SandeepaOHRM) [reply symfony/plugins/orangehrmTimePlugin/Report/EmployeeReport.php:124]: done

File: symfony/plugins/orangehrmTimePlugin/Report/EmployeeReport.php [hunk 9]
```diff
@@ -19,61 +19,208 @@
...
+                )
+            ),
+            $endpoint->getValidationDecorator()->notRequiredParamRule(
+                new ParamRule(
+                    self::FILTER_PARAMETER_ACTIVITY_ID,
+                    new Rule(Rules::POSITIVE)
```
[the remark this answers is not in the document]
Comment (SandeepaOHRM) [reply symfony/plugins/orangehrmTimePlugin/Report/EmployeeReport.php:130]: done

File: symfony/plugins/orangehrmTimePlugin/Report/EmployeeReport.php [hunk 10]
```diff
@@ -19,61 +19,208 @@
...
+                            $toDate = $endpoint->getRequestParams()->getDateTimeOrNull(
+                                RequestParams::PARAM_TYPE_QUERY,
+                                self::FILTER_PARAMETER_TO_DATE
+                            );
+
+                            if (!is_null($fromDate) && !is_null($toDate)) {
+                                if ($fromDate > $toDate) {
+                                    return false;
+                                }
+                            }
```
[the remark this answers is not in the document]
Comment (SandeepaOHRM) [reply symfony/plugins/orangehrmTimePlugin/Report/EmployeeReport.php:159]: done

File: symfony/plugins/orangehrmTimePlugin/Report/EmployeeReportData.php [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+    }
+
+    public function getMeta(): ?ParameterBag
+    {
+        $employee = $this->getEmployeeService()->getEmployeeDao()->getEmployeeByEmpNumber(
+            $this->filterParams->getEmpNumber()
+        );
+        $totalDuration = $this->getTimesheetService()->getTimesheetDao()->getTotalDurationForEmployeeReport(
+            $this->filterParams
+        );
```
[the remark this answers is not in the document]
Comment (SandeepaOHRM) [reply symfony/plugins/orangehrmTimePlugin/Report/EmployeeReportData.php:76]: done

File: symfony/plugins/orangehrmTimePlugin/Report/EmployeeReportData.php [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+
+    public function normalize(): array
+    {
+        $timesheetItems = $this->getTimesheetService()
+            ->getTimesheetDao()
+            ->getTimesheetItemsForEmployeeReport($this->filterParams);
+        $result = [];
+        foreach ($timesheetItems as $timesheetItem) {
+            $projectName = $timesheetItem[0]->getProject()->getCustomer()->getName(
+                ).' - '.$timesheetItem[0]->getProject()->getName();
```
[the remark this answers is not in the document]
Comment (SandeepaOHRM) [reply symfony/plugins/orangehrmTimePlugin/Report/EmployeeReportData.php:57]: done

File: symfony/plugins/orangehrmTimePlugin/Api/TimeReportAPI.php (hunk 6 above)
Comment (SandeepaOHRM) [reply symfony/plugins/orangehrmTimePlugin/Api/TimeReportAPI.php:31]: done

**Review by RajithaKumara — APPROVED** [review]

**Pull request merged**